app/auth/gitlab_auth.py

```diff
@@ -105,3 +108,22 @@ def logout():
         response = render_template("gitlab_logout.html", logout_url=logout_url)
 
     return response
+
+
+@blueprint.route("/exchange", methods=["GET"])
+@keycloak_authenticated
+def exchange(*, sub, access_token):
+    """Exchange a keycloak JWT for a valid Gitlab oauth token."""
+    redis_key = get_redis_key_from_token(
+        access_token, key_suffix=current_app.config["GL_SUFFIX"]
+    )
+    # NOTE: getting the client from the redis store automatically refreshes if needed
+    gl_oauth_client: RenkuWebApplicationClient = current_app.store.get_oauth_client(
+        redis_key
+    )
+    return jsonify(
+        {
+            "access_token": gl_oauth_client.access_token,
+            "expires_at": gl_oauth_client._expires_at,
```

Comment:
I know this doesn't come from you nor this PR, but why do we have the expiry date starting with `_` as if it were supposed to be a private property? Should we change it to `expires_at` in the `RenkuWebApplicationClient` and adapt to the change where it's used?

Reply:
@lorenzo-cavazzi I am hesitant to touch this because `gl_oauth_client` is an instance of `RenkuWebApplicationClient` which is a child of `oauthlib.oauth2.WebApplicationClient`. And the field `_expires_at` comes from the parent class (from the external library). And we already use `_expires_at` in several  places in our code that I did not touch at all. For example, we use it when we refresh access tokens and when we decide if a token expires soon. So I can make a parameter or function that uses `_expires_at` under the hood and returns the same value but is "public". But this is just cosmetics, our dependence on `_expires_at` is still there. And removing this dependence would require a refactor of the code that I do not want to do right now.

Reply:
The plan is to fully refactor the gateway eventually and streamline/remove a lot of these problems. I plan on submitting a RFC for this.